### ISSUE TITLE
Cannot log in netopeer2-server - ssh handshake issue

### ISSUE THREAD

**Issue by IvanAguiar95** [CESNET/Netopeer2#529]
Hello,

we are trying to install and use the Netopeer2 stack, and we'd like to connect cli and server, but we had an issue:

Clients' log:

> nc VERBOSE: Trying to connect via IPv4 to 127.0.0.1:830.
nc VERBOSE: Successfully connected to localhost:830 over IPv4.
[2020/01/14 08:22:58.914872, 2] ssh_config_parse_line:  Unapplicable option: SendEnv, line: 53
[2020/01/14 08:22:58.914885, 2] ssh_config_parse_line:  Unsupported option: HashKnownHosts, line: 54
[2020/01/14 08:22:58.914896, 2] ssh_connect:  libssh 0.8.90 (c) 2003-2019 Aris Adamantiadis, Andreas Schneider and libssh contributors. Distributed under the LGPL, please refer to COPYING file for information about your rights, using threading threads_pthread
[2020/01/14 08:22:58.914902, 2] ssh_connect:  Socket connecting, now waiting for the callbacks to work
[2020/01/14 08:22:58.914909, 1] socket_callback_connected:  Socket connection callback: 1 (0)
[2020/01/14 08:22:59.012703, 2] ssh_client_connection_callback:  SSH server banner: SSH-2.0-libssh_0.8.90
[2020/01/14 08:22:59.012783, 2] ssh_analyze_banner:  Analyzing banner: SSH-2.0-libssh_0.8.90
[2020/01/14 08:22:59.012899, 1] ssh_known_hosts_read_entries:  Failed to open the known_hosts file '/etc/ssh/ssh_known_hosts': No such file or directory
[2020/01/14 08:22:59.013015, 2] ssh_kex_select_methods:  Negotiated curve25519-sha256,ssh-rsa,aes256-gcm@openssh.com,aes256-gcm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-256-etm@openssh.com,none,none,,
[2020/01/14 08:22:59.044477, 2] ssh_init_rekey_state:  Set rekey after 4294967296 blocks
[2020/01/14 08:22:59.044545, 2] ssh_init_rekey_state:  Set rekey after 4294967296 blocks
[2020/01/14 08:22:59.044596, 2] ssh_packet_client_curve25519_reply:  SSH_MSG_NEWKEYS sent
[2020/01/14 08:22:59.046523, 2] ssh_packet_newkeys:  Received SSH_MSG_NEWKEYS
[2020/01/14 08:22:59.046645, 2] ssh_packet_newkeys:  Signature verified and valid
[2020/01/14 08:22:59.048847, 1] ssh_packet_userauth_failure:  Access denied for 'none'. Authentication that can continue: publickey,keyboard-interactive,password
[2020/01/14 08:22:59.048931, 2] ssh_packet_userauth_failure:  Access denied for 'none'. Authentication that can continue: publickey,keyboard-interactive,password
nc VERBOSE: Keyboard-interactive authentication.
Interactive SSH Authentication
Type your password:
Password: 
nc VERBOSE: Authentication successful.
[2020/01/14 08:23:00.868921, 2] channel_open:  Creating a channel 43 with 64000 window and 32768 max packet
[2020/01/14 08:23:00.870227, 2] ssh_packet_channel_open_conf:  Received a CHANNEL_OPEN_CONFIRMATION for channel 43:43
[2020/01/14 08:23:00.870258, 2] ssh_packet_channel_open_conf:  Remote window : 32000, maxpacket : 35000
[2020/01/14 08:23:00.871016, 2] grow_window:  growing window (channel 43:43) to 1280000 bytes
[2020/01/14 08:23:00.871042, 2] channel_request:  Channel request subsystem success
ly VERBOSE: Plugin "/usr/local/lib/libyang/extensions/metadata.so" successfully loaded.
ly VERBOSE: Plugin "/usr/local/lib/libyang/extensions/yangdata.so" successfully loaded.
ly VERBOSE: Plugin "/usr/local/lib/libyang/extensions/nacm.so" successfully loaded.
ly VERBOSE: Plugin "/usr/local/lib/libyang/user_types/user_inet_types.so" successfully loaded.
ly VERBOSE: Plugin "/usr/local/lib/libyang/user_types/user_yang_types.so" successfully loaded.
ly VERBOSE: Reading module "ietf-yang-metadata".
ly VERBOSE: Module "ietf-yang-metadata@2016-08-05" successfully parsed as implemented.
ly VERBOSE: Reading module "yang".
ly VERBOSE: Resolving "yang" unresolved schema nodes and their constraints...
ly VERBOSE: All "yang" schema nodes and constraints resolved.
ly VERBOSE: Module "yang@2017-02-20" successfully parsed as implemented.
ly VERBOSE: Reading module "ietf-inet-types".
ly VERBOSE: Resolving derived type "union" failed, it will be attempted later.
ly VERBOSE: Resolving derived type "union" failed, it will be attempted later.
ly VERBOSE: Resolving derived type "union" failed, it will be attempted later.
ly VERBOSE: Resolving derived type "union" failed, it will be attempted later.
ly VERBOSE: Resolving "ietf-inet-types" unresolved schema nodes and their constraints...
ly VERBOSE: All "ietf-inet-types" schema nodes and constraints resolved.
ly VERBOSE: Module "ietf-inet-types@2013-07-15" successfully parsed as implemented.
ly VERBOSE: Reading module "ietf-yang-types".
ly VERBOSE: Module "ietf-yang-types@2013-07-15" successfully parsed as implemented.
nc ERROR: Server's <hello> timeout elapsed.
cmd_connect: Connecting to the localhost:830 as user "netconf" failed.

Server's log:

> [INF]: LN: Accepted a connection on 0.0.0.0:830.
[INF]: SR: Session 4 (user "root") created.
[2020/01/14 08:16:10.285104, 2] ssh_pki_import_privkey_base64:  Trying to decode privkey passphrase=false
[2020/01/14 08:16:10.287404, 2] ssh_server_connection_callback:  SSH client banner: SSH-2.0-libssh_0.8.90
[2020/01/14 08:16:10.287433, 2] ssh_analyze_banner:  Analyzing banner: SSH-2.0-libssh_0.8.90
[2020/01/14 08:16:10.289874, 2] ssh_kex_select_methods:  Negotiated curve25519-sha256,ssh-rsa,aes256-gcm@openssh.com,aes256-gcm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-256-etm@openssh.com,none,none,,
[2020/01/14 08:16:10.309513, 2] ssh_packet_server_curve25519_init:  SSH_MSG_KEX_ECDH_REPLY sent
[2020/01/14 08:16:10.309640, 2] ssh_init_rekey_state:  Set rekey after 4294967296 blocks
[2020/01/14 08:16:10.316695, 2] ssh_init_rekey_state:  Set rekey after 4294967296 blocks
[2020/01/14 08:16:10.316753, 2] ssh_packet_server_curve25519_init:  SSH_MSG_NEWKEYS sent
[2020/01/14 08:16:10.319014, 2] ssh_packet_newkeys:  Received SSH_MSG_NEWKEYS
[INF]: LN: Received an SSH message "request-service" of subtype "ssh-userauth".
[INF]: LN: Received an SSH message "request-auth" of subtype "none".
[INF]: LN: Received an SSH message "request-auth" of subtype "interactive".
[2020/01/14 08:16:10.321270, 2] ssh_message_auth_interactive_request:  Warning: Got a keyboard-interactive response but it seems we didn't send the request.
[INF]: LN: Received an SSH message "request-auth" of subtype "interactive".
[INF]: LN: User "netconf" authenticated.
[INF]: LN: Received an SSH message "request-channel-open" of subtype "session".
[INF]: LN: Received an SSH message "request-channel" of subtype "subsystem".
[2020/01/14 08:16:12.201101, 2] channel_rcv_change_window:  Adding 1217017 bytes to channel (43:43) (from 61084 bytes)
[2020/01/14 08:16:12.201188, 2] grow_window:  growing window (channel 43:43) to 1280000 bytes
[ERR]: LN: Client's <hello> timeout elapsed.
[2020/01/14 08:16:12.201516, 1] ssh_socket_exception_callback:  Socket exception callback: 1 (0)
[2020/01/14 08:16:12.201571, 1] ssh_socket_exception_callback:  Socket error: disconnected
[INF]: LN: Accepted a connection on 0.0.0.0:830.
[INF]: SR: Session 5 (user "root") created.
[2020/01/14 08:22:59.010110, 2] ssh_pki_import_privkey_base64:  Trying to decode privkey passphrase=false
[2020/01/14 08:22:59.012551, 2] ssh_server_connection_callback:  SSH client banner: SSH-2.0-libssh_0.8.90
[2020/01/14 08:22:59.012582, 2] ssh_analyze_banner:  Analyzing banner: SSH-2.0-libssh_0.8.90
[2020/01/14 08:22:59.015404, 2] ssh_kex_select_methods:  Negotiated curve25519-sha256,ssh-rsa,aes256-gcm@openssh.com,aes256-gcm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-256-etm@openssh.com,none,none,,
[2020/01/14 08:22:59.035641, 2] ssh_packet_server_curve25519_init:  SSH_MSG_KEX_ECDH_REPLY sent
[2020/01/14 08:22:59.035745, 2] ssh_init_rekey_state:  Set rekey after 4294967296 blocks
[2020/01/14 08:22:59.035798, 2] ssh_init_rekey_state:  Set rekey after 4294967296 blocks
[2020/01/14 08:22:59.043320, 2] ssh_packet_server_curve25519_init:  SSH_MSG_NEWKEYS sent
[2020/01/14 08:22:59.046384, 2] ssh_packet_newkeys:  Received SSH_MSG_NEWKEYS
[INF]: LN: Received an SSH message "request-service" of subtype "ssh-userauth".
[INF]: LN: Received an SSH message "request-auth" of subtype "none".
[INF]: LN: Received an SSH message "request-auth" of subtype "interactive".
[2020/01/14 08:22:59.049117, 2] ssh_message_auth_interactive_request:  Warning: Got a keyboard-interactive response but it seems we didn't send the request.
[INF]: LN: Received an SSH message "request-auth" of subtype "interactive".
[INF]: LN: User "netconf" authenticated.
[INF]: LN: Received an SSH message "request-channel-open" of subtype "session".
[INF]: LN: Received an SSH message "request-channel" of subtype "subsystem".
[2020/01/14 08:23:00.873563, 2] channel_rcv_change_window:  Adding 1217017 bytes to channel (43:43) (from 61084 bytes)
[2020/01/14 08:23:00.873654, 2] grow_window:  growing window (channel 43:43) to 1280000 bytes
[ERR]: LN: Client's <hello> timeout elapsed.
[2020/01/14 08:23:00.873956, 1] ssh_socket_exception_callback:  Socket exception callback: 1 (0)
[2020/01/14 08:23:00.874012, 1] ssh_socket_exception_callback:  Socket error: disconnected

Basically, the ssh is working, but there is an error on capabilities, we guess, because <hello> is elapsing a timeout.
[ERR]: LN: Client's < hello > timeout elapsed.
and
[ERR]: LN: Server's < hello > timeout elapsed.

The versions are:
sysrepo - master
netopeer2 - master
libnetconf2 - master
libssh - we tried on master, 0.7 and stable-0.8

Any way to fix it? 

Thank you in advance.

**Comment by michalvasko** [comment]
Hi,
the log you provided is using `master` (0.8.90 version), where this issue is known, because it is also in the latest release 0.9.3. I, myself, am using a downgraded *libssh* 0.8.7 but I believe any release from at least 0.8.0 to 0.9.2 should be fine. Except maybe the latest 0.8.8 which may easily include the same regression bug.

Regards,
Michal

**Comment by IvanAguiar95** [comment]
Thank you for your help.
It works.
